include/mxnet/lib_api.h

```diff
@@ -229,20 +241,75 @@ enum MXReturnValue {
   MX_SUCCESS = 1,
 };
 
+// For sparse input, read/write the data from NDarray via pointers.
+struct MXInSparse {
+  // Pointer to data.
+  void *data{nullptr};
+  // length of (non-zero) data.
+  int64_t data_len;
+
+  // To store aux data for sparse.
+  // For CSR, indices stores the col index of non-zero values.
+  // For row sparse, indices store row index of rows which have non-zero values.
+  int64_t* indices;
+  int64_t indices_len;
+
+  // For CSR, indptr gives the start and end index of data for each row.
+  // For row sparse, indptr is empty. 
+  int64_t* indptr;
+  int64_t indptr_len;
+
+  void set(void *data_ptr, const int64_t* dims, int ndims, void *idx,
+          int64_t num_idx, void *idx_ptr = nullptr, int64_t num_idx_ptr = 0) {
+    data = data_ptr;
+    // If CSR, num of non-zero value is num_idx,
+    // If row sparse, num of value is num_idx * width.
+    data_len = idx_ptr ? num_idx : num_idx * dims[1];
```

Comment:
Does `num_idx * dims[1];` only work for 2D arrays? According to the MXNet docs:
> `indices` array stores the row index for each row slice with non-zero elements.

`indices` points to the first dimensions that have non-zero elements. Would we would need to recursively multiply the `dims` from dimension[1] to dimension[end] and multiply that by `num_idx`?
https://github.com/deepinsight/mxnet/blob/master/docs/tutorials/sparse/row_sparse.md#row-sparse-format

Reply:
You are right. Will update.